### PR TITLE
improve(subscriber_v2): flush offset commits before terminate

### LINKED ISSUE #442: Graceful shutdown
Hi guys. 
I think I'm doing something wrong, seems that I don't get the supervision strategy.
I have a pipeline that parses and sends all messages to a new topic. 
When I shut down the server, some messages are sent but not commited, which causes duplication (I need exactly once delivery).
Here is an example code (I'm sorry, it's in elixir, but could be helpful for other folks).

```elixir
defmodule TestPipeline do
  @behaviour :brod_group_subscriber_v2
  @source :test_consumer

  require Logger

  def child_spec(_opts) do
    %{
      id: __MODULE__,
      start: {__MODULE__, :start, []},
      type: :worker,
      restart: :permanent,
      shutdown: 10_000
    }
  end

  def start() do
    group_config = [
      offset_commit_policy: :commit_to_kafka_v2,
      offset_commit_interval_seconds: 2,
      rejoin_delay_seconds: 3,
      reconnect_cool_down_seconds: 10
    ]

    config = %{
      client: @source,
      group_id: "test",
      topics: ["test"],
      cb_module: __MODULE__,
      group_config: group_config,
      consumer_config: [begin_offset: :earliest]
    }

    :brod.start_link_group_subscriber_v2(config)
  end

  def init(_arg, _arg2) do
    {:ok, []}
  end

  def handle_message(messages, _state) do
    IO.inspect("processing message #{inspect(messages)}") # here I send the same batch to the new topic.
    :timer.sleep(3_000)
    IO.inspect("commiting offset")
    {:ok, :commit, nil}
  end
end
```

and my test:
```elixir
  test "shutdown gracefully" do
    brokers = [{:kafka, 9092}]

    :ok = :brod.start_client(brokers, :test_consumer)

    {:ok, pid} = start_supervised(TestPipeline)
    :timer.sleep(1_000)
    :init.stop()
    IO.inspect("shutting down")
    :timer.sleep(20_000)
  end
```

And it shows:
```
"processing message .... (always same messages)"
"shutting down"
```
But the offset is not commited

Thanks!

### PR DESCRIPTION
**Pull request by zmstone**
fixes #442

### CODE REVIEW

**Review by k32 — APPROVED** [review]

**Comment by zmstone** [comment]
tests to be added.

File: src/brod_group_subscriber_v2.erl [hunk 1]
```diff
@@ -379,14 +378,29 @@ handle_info(_Info, State) ->
...
 
+%% best-effort commits flush, this is a synced call,
+%% worst case scenario, it may timeout after 5 seconds.
+flush_offset_commits(GroupId, Coordinator) when is_pid(Coordinator) ->
+    case brod_group_coordinator:commit_offsets(Coordinator) of
+        ok -> ok;
+        {error, Reason} ->
+            ?BROD_LOG_ERROR("group_subscriber_v2 ~s failed to flush commits "
+                            "before termination", [GroupId, Reason])
+    end;
```
Comment (mikpe) [inline src/brod_group_subscriber_v2.erl:401]: Nit: looks like you have deeper indentation here than in the rest of the module.
LGTM otherwise.

**Comment by zmstone** [comment]
rebased. last two commits are new.

File: src/brod_group_subscriber_v2.erl [hunk 2]
```diff
@@ -379,28 +379,38 @@ handle_info(_Info, State) ->
...
+        ok -> ok;
+        {error, Reason} ->
+            ?BROD_LOG_ERROR("group_subscriber_v2 ~s failed to flush commits "
+                            "before termination ~p", [GroupId, Reason])
+    end;
+flush_offset_commits(_, _) ->
+    ok.
+
 -spec handle_worker_failure(brod:topic_partition(), pid(), term(), state()) ->
                                no_return().
```
Comment (mikpe) [inline src/brod_group_subscriber_v2.erl]: nit: this function no longer calls exit/1 so the -spec is a bit off

**Review by mikpe — APPROVED** [review]

File: src/brod_group_subscriber_v2.erl (hunk 2 above)
Comment (zmstone) [reply src/brod_group_subscriber_v2.erl]: thank you @mikpe 
Fixed it and squashed to the second commit.

**Pull request merged**